### PR TITLE
[WIP] VPN connections to the providers

### PR DESCRIPTION
**Pull request by mfalesni**
Currently there is no direct access to the OpenStack Undercloud internal network. I managed to install OpenVPN server to the RHOS UC node and I was able to make it work both with appliance and with testing environment. The only additional requirement to our testing setup is sudo without password so the testing suite can spin up an OpenVPN client. Rest of the things should be done automatically with proper urls in yaml.

I added my Runner package as a requirement to make it easier to run commands on local machine.

I am currently working on having the `provider_vpn` fixture included in the test parametrization.

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 02dc4ee4c3f01e3d18fc69b24adc36faba7eda62:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by seandst** [comment]
> Requirements have changed. @seandst , @psav ?
> CFME QE Bot

^
I don't think we need this anymore.

File: requirements.txt [hunk 1]
```diff
@@ -31,6 +31,7 @@ pywinrm
 PyYAML
 requests
 riggerlib
+Runner
```
Comment (seandst) [inline requirements.txt]: There are several libraries that do this, why make a new one and then not follow pep8?

**Comment by seandst** [comment]
Installing the EPEL repo pretty much invalidates test suite, since adding external the repos means we're no longer testing the distribution that we're offering for download. After the EPEL install, it must immediately be disabled and explicitly requested:

So either
`sudo yum-config-manager --disable 'epel*'`
or if we don't have yum-config-manager,
`sed -i -e "s/^enabled=1/enabled=0/" /etc/yum.repos.d/epel.repo`

and then
`yum --enablerepo=epel -y install openvpn`

File: utils/appliance.py [hunk 2]
```diff
@@ -1116,6 +1117,120 @@ def wait_for_web_ui(self, timeout=900, running=True, log_callback=None):
             fail_condition=not running, delay=10)
         return result
 
+    def install_epel(self, log_callback=None):
+        """Install the EPEL repository on the appliance"""
+        if log_callback is None:
+            log_callback = self.log.info
+
+        def log_raise(exception_class, message):
```
Comment (seandst) [inline utils/appliance.py]: This defined in two places, and should probably be a function in the module.

File: utils/appliance.py [hunk 3]
```diff
@@ -1116,6 +1117,120 @@ def wait_for_web_ui(self, timeout=900, running=True, log_callback=None):
...
+                log_callback("OpenVPN not set up")
+                return
+            client.run_command("service openvpn stop")
+            assert client.run_command("chkconfig openvpn off").rc == 0
+            files = client.run_command("cat /ovpn_list").output.strip()
+            for file in [file.strip() for file in files.split("\n")]:
+                client.run_command("rm -f {}".format(file))
+            client.run_command("rm -f /ovpn_list")
+
+    def uninstall_epel(self, log_callback=None):
```
Comment (seandst) [inline utils/appliance.py]: If EPEL is installed and disabled, as it should be, we shouldn't need to ever uninstall it.

File: utils/local_vpn.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+@contextmanager
+def vpn_for(provider_key):
+    provider_data = conf.cfme_data.management_systems[provider_key]
+    if "vpn" not in provider_data:
+        yield
+    else:
+        assert Run.command("rpm -qa | grep ^openvpn", shell=True), "OpenVPN not installed locally!"
+        tmpdir = "/tmp/{}".format(generate_random_string())
+        logger.info("OpenVPN temporary directory: {}".format(tmpdir))
+        assert Run.command("mkdir -p {}".format(tmpdir))
```
Comment (seandst) [inline utils/local_vpn.py]: Do not use asserts for program flow.

File: utils/local_vpn.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        assert Run.command(
+            "wget -O {}/tmp.tar.gz {}".format(tmpdir, provider_data["vpn"]["test_machine"]))
+        assert Run.command("cd {}; tar xfv tmp.tar.gz".format(tmpdir), shell=True)
+        log_filename = "openvpn-{}.log".format(generate_random_string())
+        logger.info("OpenVPN log: {}".format(log_filename))
+        with path.log_path.join(log_filename).open("w") as log:
+            logger.info("OpenVPN starting!")
+            openvpn = Popen(
+                "cd {} && exec sudo -n openvpn --config user.conf".format(tmpdir),
```
Comment (seandst) [inline utils/local_vpn.py]: The test runner will never ever run as root or have sudo privs.

**Pull request closed without merging**